### PR TITLE
Semantic tokens

### PR DESCRIPTION
**Pull request by Kha**
Highlight user-defined keywords just like buil-in ones using LSP's semantic tokens requests
image

TODO:
- [x] I can't seem to get it to work in VS Code even after setting `editor.semanticHighlighting.enabled`. Is there anything else that needs to be done?
- [x] Not sure why it ignores quotations
- [x] Add at least one more helpful token kind, e.g. for field notation

### CODE REVIEW

**Comment by gebner** [comment]
> I can't seem to get it to work in VS Code

For me this works out of the box, though the color is off:
semantic_highlighting

**Comment by gebner** [comment]
It turns out you can customize how the semantic token selectors get mapped to the scopes:
semantic_highlighting2

Small bug in `unifHintAndTC.lean`:
semhibug

**Comment by gebner** [comment]
Another regression is that sorry is now highlighted as a normal keyword.

**Comment by gebner** [comment]
The first line of docstrings is apparently also a keyword.

**Comment by Kha** [comment]
Thanks for the feedback! It works for me as well now after updating to 0.0.12.

File: src/Lean/Server/AsyncList.lean [hunk 1]
```diff
@@ -70,17 +70,18 @@ partial def getAll : AsyncList ε α → List α × Option ε
     | Except.ok tl => tl.getAll
     | Except.error e => ⟨[], some e⟩
 
-/-- Spawns a `Task` waiting on which ensures that the async tail, if it exists,
-is fully executed. This is useful if the computation has side effects.
-The task returns the same values as `getAll`. -/
-partial def waitAll [Coe Error ε] : AsyncList ε α → IO (Task (List α × Option ε))
+/-- Spawns a `Task` waiting on the prefix of elements for which `p` is true. -/
```
Comment (Vtec234) [inline src/Lean/Server/AsyncList.lean:73]: Why remove the docstring?

File: src/Lean/Server/FileWorker.lean [hunk 2]
```diff
@@ -528,6 +528,45 @@ section RequestHandling
...
+      for s in snaps do
+        if s.endPos > beginPos then
+          for stx in s.stx.topDown (firstChoiceOnly := true) do
+            -- highlight keywords
+            if let Syntax.atom info val := stx then
+              if let some pos := info.getPos? then
+                if val.bsize > 0 && val[0].isAlpha && beginPos <= pos && pos < endPos then
+                  let lspPos' := text.utf8PosToLspPos pos
+                  let deltaLine := lspPos'.line - lspPos.line
+                  let deltaStart := lspPos'.character - (if lspPos'.line == lspPos.line then lspPos'.character else 0)
```
Comment (Vtec234) [inline src/Lean/Server/FileWorker.lean:549]: Should maybe subtract `lspPos.character` here for same-line relative start positions?

File: src/Lean/Data/Lsp/LanguageFeatures.lean [hunk 3]
```diff
@@ -140,5 +140,79 @@ structure DocumentSymbolResult where
...
+
+structure SemanticTokensLegend where
+  tokenTypes : Array String
+  tokenModifiers : Array String
+  deriving FromJson, ToJson
+
+structure SemanticTokensOptions where
+  legend : SemanticTokensLegend
+  range : Bool
+  full : Bool /- | {
```
Comment (Vtec234) [inline src/Lean/Data/Lsp/LanguageFeatures.lean:198]: Not sure I'm reading the spec right but it seems to mean we should have `Either Bool SemanticTokensOptionsDelta` here with the right variant being `structure SemanticTokensOptionsDelta where delta?: Bool` ?

File: src/Lean/Server/FileWorker.lean (hunk 2 above)
Comment (Kha) [reply src/Lean/Server/FileWorker.lean]: Good catch, this resolves the quotation and other issues mentioned above!

File: src/Lean/Data/Lsp/LanguageFeatures.lean (hunk 3 above)
Comment (Kha) [reply src/Lean/Data/Lsp/LanguageFeatures.lean:198]: This is part of the server capabilities, so no need to support a variant we don't use

**Comment by gebner** [comment]
When browsing through the Lean source code (opening random files), I often get the following error:
```
Watchdog error: Got unknown document URI (file:///home/gebner/lean4/src/Init/Coe.lean)
[Info  - 11:42:10 AM] Connection to server got closed. Server will restart.
[Error - 11:42:10 AM] Request textDocument/semanticTokens/full failed.
Error: Connection got disposed.
	at Object.dispose (/home/gebner/.vscode-oss/extensions/leanprover.lean4-0.0.12/out/extension.js:2:145530)
	at Object.dispose (/home/gebner/.vscode-oss/extensions/leanprover.lean4-0.0.12/out/extension.js:2:225702)
	at C.handleConnectionClosed (/home/gebner/.vscode-oss/extensions/leanprover.lean4-0.0.12/out/extension.js:2:225915)
	at C.handleConnectionClosed (/home/gebner/.vscode-oss/extensions/leanprover.lean4-0.0.12/out/extension.js:2:286303)
	at t (/home/gebner/.vscode-oss/extensions/leanprover.lean4-0.0.12/out/extension.js:2:224004)
	at invoke (/home/gebner/.vscode-oss/extensions/leanprover.lean4-0.0.12/out/extension.js:2:147143)
	at o.fire (/home/gebner/.vscode-oss/extensions/leanprover.lean4-0.0.12/out/extension.js:2:147904)
	at Y (/home/gebner/.vscode-oss/extensions/leanprover.lean4-0.0.12/out/extension.js:2:134788)
	at invoke (/home/gebner/.vscode-oss/extensions/leanprover.lean4-0.0.12/out/extension.js:2:147143)
	at o.fire (/home/gebner/.vscode-oss/extensions/leanprover.lean4-0.0.12/out/extension.js:2:147904)
	at m.fireClose (/home/gebner/.vscode-oss/extensions/leanprover.lean4-0.0.12/out/extension.js:2:159357)
	at Socket.<anonymous> (/home/gebner/.vscode-oss/extensions/leanprover.lean4-0.0.12/out/extension.js:2:160142)
	at Socket.emit (events.js:315:20)
	at Pipe.<anonymous> (net.js:674:12)
```

**Comment by Kha** [comment]
Heh, I assume this is because we violate the following part of the spec:

> "Note that a server’s ability to fulfill requests is independent of whether a text document is open or closed."

I'm pretty sure we don't want to change that, but we should probably convert this case to a stderr warning (and return `Inhabited.default`...?)

**Comment by Kha** [comment]
Now with highlighting for parameters/variables and projections

image

**Comment by Kha** [comment]
There are probably dozen of further cases to be added or suppressed, but I think this is stable enough now to be merged

**Pull request merged**